File: j2kt/jre/java/common/java/util/Comparator.java
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package java.util;

import static javaemul.internal.KtNativeUtils.ktNative;

import java.util.function.Function;
import java.util.function.ToDoubleFunction;
import java.util.function.ToIntFunction;
import java.util.function.ToLongFunction;
import javaemul.internal.annotations.KtNative;
import org.jspecify.nullness.NullMarked;
import org.jspecify.nullness.Nullable;

/**
 * An interface used a basis for implementing custom ordering. <a
 * href="http://java.sun.com/j2se/1.5.0/docs/api/java/util/Comparator.html">[Sun docs]</a>
 *
 * @param <T> the type to be compared.
 */
@KtNative(
    value = "kotlin.Comparator",
    bridgeWith = "java.util.Comparator",
    companionObject = "java.util.Comparator")
@FunctionalInterface
@NullMarked
public interface Comparator<T extends @Nullable Object> {

  int compare(T a, T b);

  default Comparator<T> reversed() {
    return ktNative();
  }

  default Comparator<T> thenComparing(Comparator<? super T> other) {
    return ktNative();
  }

  default <U extends @Nullable Object> Comparator<T> thenComparing(
      Function<? super T, ? extends U> keyExtractor, Comparator<? super U> keyComparator) {
    return ktNative();
  }

  default <U extends Comparable<? super U>> Comparator<T> thenComparing(
      Function<? super T, ? extends U> keyExtractor) {
    return ktNative();
  }

  default Comparator<T> thenComparingInt(ToIntFunction<? super T> keyExtractor) {
    return ktNative();
  }

  default Comparator<T> thenComparingLong(ToLongFunction<? super T> keyExtractor) {
    return ktNative();
  }

  default Comparator<T> thenComparingDouble(ToDoubleFunction<? super T> keyExtractor) {
    return ktNative();
  }

  static <T extends @Nullable Object, U extends @Nullable Object> Comparator<T> comparing(
      Function<? super T, ? extends U> keyExtractor, Comparator<? super U> keyComparator) {
    return ktNative();
  }

  static <T extends @Nullable Object, U extends Comparable<? super U>> Comparator<T> comparing(
      Function<? super T, ? extends U> keyExtractor) {
    return ktNative();
  }

  static <T extends @Nullable Object> Comparator<T> comparingDouble(
      ToDoubleFunction<? super T> keyExtractor) {
    return ktNative();
  }

  static <T extends @Nullable Object> Comparator<T> comparingInt(
      ToIntFunction<? super T> keyExtractor) {
    return ktNative();
  }

  static <T extends @Nullable Object> Comparator<T> comparingLong(
      ToLongFunction<? super T> keyExtractor) {
    return ktNative();
  }

  static <T extends Comparable<? super T>> Comparator<T> naturalOrder() {
    return ktNative();
  }

  static <T> Comparator<@Nullable T> nullsFirst(Comparator<? super T> comparator) {
    return ktNative();
  }

  static <T> Comparator<@Nullable T> nullsLast(Comparator<? super T> comparator) {
    return ktNative();
  }

  static <T extends Comparable<? super T>> Comparator<T> reverseOrder() {
    return ktNative();
  }
}

File: j2kt/jre/java/native/java/util/Comparator.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package java.util

import java.util.function.Function
import java.util.function.ToDoubleFunction
import java.util.function.ToIntFunction
import java.util.function.ToLongFunction
import kotlin.comparisons.reversed as default_reversed

fun interface Comparator<T> : kotlin.Comparator<T> {
  override fun compare(a: T, b: T): Int

  fun reversed(): kotlin.Comparator<T> = default_reversed()

  fun thenComparing(other: kotlin.Comparator<in T>): kotlin.Comparator<T> = then(other)

  fun <U> thenComparing(
    keyExtractor: Function<in T, out U>,
    keyComparator: kotlin.Comparator<in U>
  ): kotlin.Comparator<T> = thenBy(keyComparator, keyExtractor::apply)

  fun <U : Comparable<U>> thenComparing(keyExtractor: Function<in T, out U>): kotlin.Comparator<T> =
    thenBy(keyExtractor::apply)

  fun thenComparingInt(keyExtractor: ToIntFunction<in T>): kotlin.Comparator<T> =
    then(comparingInt(keyExtractor))

  fun thenComparingLong(keyExtractor: ToLongFunction<in T>): kotlin.Comparator<T> =
    then(comparingLong(keyExtractor))

  fun thenComparingDouble(keyExtractor: ToDoubleFunction<in T>): kotlin.Comparator<T> =
    then(comparingDouble(keyExtractor))

  companion object {
    fun <T, U> comparing(
      keyExtractor: Function<in T, out U>,
      keyComparator: kotlin.Comparator<in U>
    ): kotlin.Comparator<T> = compareBy(keyComparator, keyExtractor::apply)

    fun <T, U : Comparable<U>> comparing(
      keyExtractor: Function<in T, out U>
    ): kotlin.Comparator<T> = compareBy(naturalOrder(), keyExtractor::apply)

    fun <T> comparingDouble(keyExtractor: ToDoubleFunction<in T>): kotlin.Comparator<T> =
      Comparator<T> { a, b ->
        keyExtractor.applyAsDouble(a).compareTo(keyExtractor.applyAsDouble(b))
      }

    fun <T> comparingInt(keyExtractor: ToIntFunction<in T>): kotlin.Comparator<T> =
      Comparator<T> { a, b -> keyExtractor.applyAsInt(a).compareTo(keyExtractor.applyAsInt(b)) }

    fun <T> comparingLong(keyExtractor: ToLongFunction<in T>): kotlin.Comparator<T> =
      Comparator<T> { a, b -> keyExtractor.applyAsLong(a).compareTo(keyExtractor.applyAsLong(b)) }

    fun <T> nullsFirst(comparator: kotlin.Comparator<in T>): kotlin.Comparator<T?> =
      kotlin.comparisons.nullsFirst(comparator)

    fun <T> nullsLast(comparator: kotlin.Comparator<in T>): kotlin.Comparator<T?> =
      kotlin.comparisons.nullsLast(comparator)

    fun <T : Comparable<T>> naturalOrder(): kotlin.Comparator<T> = kotlin.comparisons.naturalOrder()

    fun <T : Comparable<T>> reverseOrder(): kotlin.Comparator<T> = kotlin.comparisons.reverseOrder()
  }
}
